go-controller/pkg/cluster/ha_master.go

```diff
@@ -0,0 +1,562 @@
+package cluster
+
+import (
+	"context"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/config"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+	"github.com/sirupsen/logrus"
+	kapi "k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/leaderelection"
+	"k8s.io/client-go/tools/leaderelection/resourcelock"
+)
+
+const (
+	invalidIPAddress = "0.0.0.1"
+	haLeaderLockName = "ovn-kubernetes-master"
+	ovnkubeDbEp      = "ovnkube-db"
+)
+
+//TODO: Need to make this configurable.
+const (
+	ovnNorthDbPort int32 = 6641
+	ovnSouthDbPort int32 = 6642
+)
+
+//TODO: Need to come up with proper default values.
+const (
+	leaseDuration = 6 * time.Second
+	renewDeadline = 3 * time.Second
+	retryPeriod   = 2 * time.Second
+)
+
+// OvnHAMasterController is the object holder for managing the HA master
+// cluster
+type OvnHAMasterController struct {
+	kubeClient      kubernetes.Interface
+	cluster         *OvnClusterController
+	ovnController   *ovn.Controller
+	nodeName        string
+	manageDBServers bool
+	isLeader        bool
+	leaderElector   *leaderelection.LeaderElector
+}
+
+// NewHAMasterController creates a new HA Master controller
+func NewHAMasterController(kubeClient kubernetes.Interface, cluster *OvnClusterController,
+	nodeName string, manageDBServers bool) *OvnHAMasterController {
+	ovnController := ovn.NewOvnController(kubeClient, cluster.watchFactory)
+	return &OvnHAMasterController{
+		kubeClient:      kubeClient,
+		cluster:         cluster,
+		ovnController:   ovnController,
+		nodeName:        nodeName,
+		manageDBServers: manageDBServers,
+		isLeader:        false,
+		leaderElector:   nil,
+	}
+}
+
+// StartHAMasterCluster runs the replication controller
+func (hacluster *OvnHAMasterController) StartHAMasterCluster() error {
+	if hacluster.manageDBServers {
+		// Always demote the OVN DBs to backmode mode.
+		// After the leader election, the leader will promote the OVN Dbs
+		// to become active.
+		err := hacluster.DemoteOVNDbs(invalidIPAddress, "6641", "6642")
```

Comment:
@numansiddique should we use the port number constants here?

Reply:
Done